### PR TITLE
Output coverage report direct to PR

### PR DESCRIPTION
**Pull request by alanmcruickshank**
Coveralls is being buggy again.

This posts the coverage report from out internal coverage report directly onto the PR using a comment. Tested on #5177.

### CODE REVIEW

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -358,14 +358,81 @@ jobs:
...
+            const comments = await github.paginate(
+              github.rest.issues.listComments,
+              {
+                owner: 'sqlfluff',
+                repo: 'sqlfluff',
+                issue_number: context.issue.number
+              }
+            )
+
+            comment_id = null
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:398]: I slightly prefer `let comment_id = null`

File: .github/workflows/ci-tests.yml [hunk 2]
```diff
@@ -358,14 +358,81 @@ jobs:
...
+            for (const comment of comments) {
+              if (comment.body.indexOf(trigger_str) >= 0) {
+                console.log("Found target comment ID: %d", comment.id)
+                comment_id = comment.id
+              } else {
+                console.log("Comment ID %d not valid with body:\n%s.", comment.id, comment.body)
+              }
+            }
+
+            previous_outcome = "${{ steps.report_coverage.outcome }}"
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:410]: I slightly prefer `const previous_outcome`

File: .github/workflows/ci-tests.yml [hunk 3]
```diff
@@ -358,14 +358,81 @@ jobs:
...
+            console.log("Previous coverage step outcome: %s", previous_outcome)
+            if (previous_outcome == "success") {
+              status_emoji = "✅"
+            } else {
+              status_emoji = "⚠️"
+            }
+
+            const { promises: fs } = require('fs')
+            let content = await fs.readFile('coverage-report.txt', 'utf8')
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:419]: `content` could be a `const`

**Review by greg-finley — APPROVED** [review]
Nice 👍 

Left some JS nits 😉 . Ideally would recommend to end all lines with semicolons also.

**Comment by alanmcruickshank** [comment]
> > Nice 👍
> > Left some JS nits 😉 . Ideally would recommend to end all lines with semicolons also.
> 
> Since it works and maybe is unlikely to change, I would say feel free to ignore JS nits though if you prefer

They're all good suggestions and my JS-Fu is poor. I will take all the advice I can get!

File: .github/workflows/ci-tests.yml [hunk 4]
```diff
@@ -358,14 +358,81 @@ jobs:
...
+            console.log("Got %d comments", comments.length)
+
+            for (const comment of comments) {
+              if (comment.body.indexOf(trigger_str) >= 0) {
+                console.log("Found target comment ID: %d", comment.id)
+                comment_id = comment.id
+              } else {
+                console.log("Comment ID %d not valid with body:\n%s.", comment.id, comment.body)
+              }
+            }
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:408]: Fine, one more way that's slightly more JS-thonic

```js
comments.forEach(comment => {
  if (comment.body.indexOf(trigger_str) >= 0) {
    console.log("Found target comment ID: %d", comment.id);
    comment_id = comment.id;
  } else {
    console.log("Comment ID %d not valid with body:\n%s.", comment.id, comment.body);
  }
});
```

File: .github/workflows/ci-tests.yml (hunk 3 above)
Comment (alanmcruickshank) [reply .github/workflows/ci-tests.yml]: On this one, I'm not sure I really get the impact of `let`, `const` and perhaps their combination. When should I use each?

Comment (greg-finley) [reply .github/workflows/ci-tests.yml]: `foo = true` could work for either setting or changing a variable, but `let foo = true` is saying that this is the first time we are setting this (and might help you find accidental variable shadowing)

`const foo = true` means `foo` cannot change later

File: .github/workflows/ci-tests.yml [hunk 5]
```diff
@@ -358,14 +358,84 @@ jobs:
...
+        if: always()
+        with:
+          script: |
+            // First list the existing comments
+
+            const trigger_str = 'Coverage Results';
+            console.log("Getting existing comments...");
+
+            const issue_number = context.issue.number;
+            console.log("Issue number: " + issue_number);
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:390]: @alanmcruickshank it's failing because
`Issue number: undefined`

Comment (greg-finley) [reply .github/workflows/ci-tests.yml:390]: Here's one suggestion, but I feel it wouldn't necessarily pick the right issue number: https://github.com/actions/github-script/issues/203#issuecomment-1027833024

Comment (greg-finley) [reply .github/workflows/ci-tests.yml:390]: We should probably do this? https://github.com/actions/github-script/issues/203#issuecomment-1034195974

**Pull request merged**